include/deal.II/rol/vector_adaptor.h

```diff
@@ -0,0 +1,478 @@
+    :
+    vector_ptr (vector_ptr)
+  {
+    Assert ((std::is_convertible<real_type, value_type>::value),
+            ExcMessage("The real_type of the current VectorType is not "
+                       "convertible to the value_type."));
+  }
+
+
+
+  template<typename VectorType>
+  Teuchos::RCP<VectorType>
+  VectorAdaptor<VectorType>::getVector ()
+  {
+    return vector_ptr;
+  }
+
+
+
+  template<typename VectorType>
+  Teuchos::RCP<const VectorType>
+  VectorAdaptor<VectorType>::getVector () const
+  {
+    return vector_ptr;
+  }
+
+
+
+  template<typename VectorType>
+  void
+  VectorAdaptor<VectorType>::set (const ROL::Vector<value_type> &rol_vector)
+  {
+    const VectorAdaptor &vector_adaptor =
+      Teuchos::dyn_cast<const VectorAdaptor>(rol_vector);
+
+    (*vector_ptr) =  *(vector_adaptor.getVector());
+  }
+
+
+
+  template<typename VectorType>
+  void
+  VectorAdaptor<VectorType>::plus (const ROL::Vector<value_type> &rol_vector)
+  {
+    Assert (this->dimension() == rol_vector.dimension(),
+            ExcDimensionMismatch(this->dimension(), rol_vector.dimension()));
+
+    const VectorAdaptor &vector_adaptor =
+      Teuchos::dyn_cast<const VectorAdaptor>(rol_vector);
+
+    *vector_ptr += *(vector_adaptor.getVector());
+  }
+
+
+
+  template<typename VectorType>
+  void
+  VectorAdaptor<VectorType>::axpy (const value_type               alpha,
+                                   const ROL::Vector<value_type> &rol_vector)
+  {
+    Assert (this->dimension() == rol_vector.dimension(),
+            ExcDimensionMismatch(this->dimension(), rol_vector.dimension()));
+
+    const VectorAdaptor &vector_adaptor =
+      Teuchos::dyn_cast<const VectorAdaptor>(rol_vector);
+
+    vector_ptr->add (alpha, *(vector_adaptor.getVector()));
+  }
+
+
+
+  template<typename VectorType>
+  int
+  VectorAdaptor<VectorType>::dimension () const
+  {
+    Assert (vector_ptr->size() < std::numeric_limits<int>::max(),
+            ExcMessage("The size of the vector being used is greater than "
+                       "largest value of type int."));
+    return static_cast<int>(vector_ptr->size());
+  }
+
+
+
+  template<typename VectorType>
+  void
+  VectorAdaptor<VectorType>::scale (const value_type alpha)
+  {
+    (*vector_ptr) *= alpha;
+  }
+
+
+
+  template<typename VectorType>
+  typename VectorType::value_type
+  VectorAdaptor<VectorType>::
+  dot (const ROL::Vector<value_type> &rol_vector) const
+  {
+    Assert (this->dimension() == rol_vector.dimension(),
+            ExcDimensionMismatch(this->dimension(), rol_vector.dimension()));
+
+    const VectorAdaptor &vector_adaptor =
+      Teuchos::dyn_cast< const VectorAdaptor>(rol_vector);
+
+    return (*vector_ptr) * (*vector_adaptor.getVector());
+  }
+
+
+
+  template<typename VectorType>
+  typename VectorType::value_type
+  VectorAdaptor<VectorType>::norm() const
+  {
+    return static_cast<value_type>(vector_ptr->l2_norm());
```

Comment:
This is a suggestion rather than a ""must do":
Here you could also use `internal::NumberType<value_type>::value(...)` found in `base/number.h`. Source [here](http://dealii.org/developer/doxygen/deal.II/numbers_8h_source.html#l00428).